adcc/adc_pp/state2state_transition_dm.py

```diff
@@ -20,12 +20,131 @@
 
 import libadcc
 
-DISPATCH = {}  # None implemented
+
+# TODO: just a sketch for what I'd like to have, because we need to check on
+# 2 amplitude vectors simultaneously
+def _check_have_singles_block(*amplitudes):
+    if any("s" not in amplitude.blocks for amplitude in amplitudes):
+        raise ValueError("state2state_transition_dm at ADC(0) level and "
+                         "beyond expects an excitation amplitude with a "
+                         "singles part.")
+
+
+def _check_have_doubles_block(*amplitudes):
+    if any("d" not in amplitude.blocks for amplitude in amplitudes):
+        raise ValueError("state2state_transition_dm at ADC(2) level and "
+                         "beyond expects an excitation amplitude with a "
+                         "singles and a doubles part.")
+
+
+def _check_singles_subspaces(*amplitudes):
+    for amplitude in amplitudes:
+        ul1 = amplitude["s"]
+        if ul1.subspaces != [b.o, b.v]:
+            raise ValueError("Mismatch in subspaces singles part "
+                             f"(== {ul1.subspaces}), where {b.o}{b.v} "
+                             "was expected.")
+
+
+def _check_doubles_subspaces(*amplitudes):
+    for amplitude in amplitudes:
+        ul2 = amplitude["d"]
+        if ul2.subspaces != [b.o, b.o, b.v, b.v]:
+            raise ValueError("Mismatch in subspaces doubles part "
+                             f"(== {ul2.subspaces}), where "
+                             f"{b.o}{b.o}{b.v}{b.v} was expected.")
+
+
+def s2s_tdm_adc0(mp, amplitude_l, amplitude_r, intermediates):
+    _check_have_singles_block(amplitude_l, amplitude_r)
+    _check_singles_subspaces(amplitude_l, amplitude_r)
+    ul1 = amplitude_l["s"]
+    ur1 = amplitude_r["s"]
+
+    dm = OneParticleOperator(mp, is_symmetric=False)
+    dm[b.oo] = -einsum('ja,ia->ij', ul1, ur1)
+    dm[b.vv] = einsum('ia,ib->ab', ul1, ur1)
+    return dm
+
+
+def s2s_tdm_adc2(mp, amplitude_l, amplitude_r, intermediates):
+    _check_have_doubles_block(amplitude_l, amplitude_r)
+    _check_doubles_subspaces(amplitude_l, amplitude_r)
+    dm = s2s_tdm_adc0(mp, amplitude_l, amplitude_r, intermediates)
+
+    ul1 = amplitude_l["s"]
+    ur1 = amplitude_r["s"]
+    ul2 = amplitude_l["d"]
+    ur2 = amplitude_r["d"]
+
+    t2 = mp.t2(b.oovv)
+    p0_ov = mp.mp2_diffdm[b.ov]
+    p0_oo = mp.mp2_diffdm[b.oo]
+    p0_vv = mp.mp2_diffdm[b.vv]
+    p1_oo = dm[b.oo].evaluate()  # ADC(1) tdm
+    p1_vv = dm[b.vv].evaluate()  # ADC(1) tdm
+
+    # ADC(2) ISR intermediate (TODO Move to intermediates)
+    rul1 = einsum('ijab,jb->ia', t2, ul1).evaluate()
+    rur1 = einsum('ijab,jb->ia', t2, ur1).evaluate()
+
+    dm[b.oo] = (
+        p1_oo - 2.0 * einsum('ikab,jkab->ij', ur2, ul2)
+        + 0.5 * einsum('ik,kj->ij', p1_oo, p0_oo)
+        + 0.5 * einsum('ik,kj->ij', p0_oo, p1_oo)
+        - 0.5 * einsum('ikcd,lk,jlcd->ij', t2, p1_oo, t2)
+        + einsum('ikcd,jkcb,db->ij', t2, t2, p1_vv)
```

Comment:
Again a question of style, but I think I would put a `1.0 * ` in such places.

Reply:
Done.